nvflare/client/ipc_agent.py

```diff
@@ -362,85 +348,100 @@ def submit_result(self, result: defs.TaskResult) -> bool:
         meta = result.meta
         rc = result.return_code
         data = result.data
 
         msg = new_message(
             headers={
-                defs.MsgHeader.TASK_NAME: current_task.task_name,
-                defs.MsgHeader.TASK_ID: current_task.task_id,
+                defs.MsgHeader.TASK_NAME: current_task.task_name if current_task else "",
+                defs.MsgHeader.TASK_ID: current_task.task_id if current_task else "",
                 defs.MsgHeader.RC: rc,
             },
             payload={
                 defs.PayloadKey.META: meta,
                 defs.PayloadKey.DATA: data,
             },
         )
+
+        last_send_time = 0
         while True:
-            if current_task.already_received:
+            if self.is_done or self.is_stopped:
+                self.logger.error(f"quit submitting result for task {current_task} since agent is closed")
+                raise defs.AgentClosed("agent is stopped")
+
+            if current_task and current_task.already_received:
                 if not current_task.last_send_result_time:
                     self.logger.warning(f"task {current_task} was marked already_received but has been sent!")
                 return True
 
-            if self.is_done or self.is_stopped:
-                self.logger.error(f"quit submitting result for task {current_task} since agent is closed")
-                return False
-
-            if current_task.aborted:
+            if current_task and current_task.aborted:
                 self.logger.error(f"quit submitting result for task {current_task} since it is aborted")
                 return False
 
-            current_task.last_send_result_time = time.time()
-            self.logger.info(f"sending result to {current_task.sender} for task {current_task}")
-            reply = self.cell.send_request(
-                channel=defs.CHANNEL,
-                topic=defs.TOPIC_SUBMIT_RESULT,
-                target=current_task.sender,
-                request=msg,
-                timeout=self.submit_result_timeout,
-            )
-            if reply:
-                rc = reply.get_header(MessageHeaderKey.RETURN_CODE)
-                sender = reply.get_header(MessageHeaderKey.ORIGIN)
-                if rc == ReturnCode.OK:
-                    return True
-                elif rc == ReturnCode.INVALID_REQUEST:
-                    self.logger.error(f"received return code from {sender}: {rc}")
-                    return False
-                else:
-                    self.logger.info(f"failed to send to {current_task.sender}: {rc} - will retry")
-            time.sleep(2.0)
+            if self.is_connected and time.time() - last_send_time > self.resend_result_interval:
+                self.logger.info(f"sending result to {self.peer_fqcn} for task {current_task}")
+                if current_task:
+                    current_task.last_send_result_time = time.time()
```

Comment:
if current_task is None, do we still want to do the following send_request?

Reply:
Yes. This is to support the case that the trainer crashed and then restarted. In this case, the current_task is None, but we must send the re-trained result to IPCExchanger, in case it is still waiting for the result. If the IPCExchanger is not waiting for a result, it will ignore it.